---
title: "901-CRRB"
author: "Mac Campbell"
date: "3/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#O. m. gairdneri
Does single-end RADseq help us out here?     
Created a test14.bamlist    

```{sh, eval=FALSE}
angsd -P 4 -b bamlists/test14.bamlist -minInd 39  -out outputs/900/test14-maf05p-sites -minMaf 0.05 -minMapQ 30 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 -doGeno 4 -doPost 1 -postCutoff 0.95 -doVcf 1 -rf homoblocks/non-tetrasomic-non-inversion.txt > outputs/900/test14-make-vcf.out 2> outputs/900/test14-make-vcf.err

#Sort vcf
gunzip -c outputs/900/test14-maf05p-sites.vcf.gz | awk '$1 ~ /^#/ {print $0;next} {print $0 | "sort -k1,1 -k2,2n"}' > outputs/900/test14-maf05p-sites.sort.vcf
#Compress
bgzip outputs/900/test14-maf05p-sites.sort.vcf 
#Index
tabix outputs/900/test14-maf05p-sites.sort.vcf.gz -f

#Pruned - as input for multispecies coalescent type analyses
bcftools +prune -l 0.9 -w 30000 outputs/900/test14-maf05p-sites.sort.vcf.gz -Ov -o outputs/900/test14-maf05p-sites.sort.prune.vcf

#lets reheader the bam file
bcftools reheader --samples bamlists/test14-sample-names.txt -o outputs/900/test14-maf05p-sites.sort.prune.reheadered.vcf outputs/900/test14-maf05p-sites.sort.prune.vcf

#Convert to phylip
./vcf2phylip.py -i outputs/900/test14-maf05p-sites.sort.prune.reheadered.vcf

#Correct for ascertainment bias
source activate py3; ./103-remove-invariant.py -p outputs/900/test14-maf05p-sites.sort.prune.reheadered.min4.phy -o outputs/900/test14-maf05p-sites.sort.prune.reheadered.min4.asc.phy; source deactivate;

#Infer a proper tree, with model testing
iqtree -s outputs/900/test14-maf05p-sites.sort.prune.reheadered.min4.asc.phy -st DNA -m MFP+ASC -bb 1000 -alrt 1000 --redo

#Unpruned
bcftools reheader --samples bamlists/test14-sample-names.txt -o outputs/900/test14-maf05p-sites.sort.reheadered.vcf.gz outputs/900/test14-maf05p-sites.sort.vcf.gz

gunzip outputs/900/test14-maf05p-sites.sort.reheadered.vcf.gz

./vcf2phylip.py -i outputs/900/test14-maf05p-sites.sort.reheadered.vcf

source activate py3; ./103-remove-invariant.py -p outputs/900/test14-maf05p-sites.sort.reheadered.min4.phy -o outputs/900/test14-maf05p-sites.sort.reheadered.min4.asc.phy; source deactivate;

iqtree -s outputs/900/test14-maf05p-sites.sort.reheadered.min4.asc.phy -st DNA -m MFP+ASC  -bb 1000 -alrt 1000 --redo



```

I end up with 41 taxa and 2199 unlinked SNPs.

How do our trees look?    

```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(ggtree)
library(ape)

tree<-read.tree(file="Data Supplement/ML tree results/ml.contree")
tree<-root(tree, c("PCT_PPP1_01","LCT_GACR_10"))
ggtree(tree)+geom_tiplab(size=2)+geom_nodelab()



```
### Generate figure

```{r}
library("phangorn")
tree<-read.tree(file="Data Supplement/ML tree results/ml.treefile")
tree<-midpoint(tree)
ggtree(tree)
```

```{r}
nodes<-tibble(tree$node.label)
nodes[1,]<-"NA"
nodes<-separate(nodes, `tree$node.label`, sep="/", into = c("SH-aLRT","UF-Boot"))
nodes$`SH-aLRT`<-as.numeric(nodes$`SH-aLRT`)
nodes$`UF-Boot`<-as.numeric(nodes$`UF-Boot`)
supported<-nodes
supported$`SH-aLRT` <-ifelse(nodes$`SH-aLRT` < 80, "NA", nodes$`SH-aLRT`)
supported$`UF-Boot` <-ifelse(nodes$`UF-Boot` < 95, "NA", nodes$`UF-Boot`)
supported$`SH-aLRT`<-as.numeric(supported$`SH-aLRT`)
supported$`UF-Boot`<-as.numeric(supported$`UF-Boot`)
ggtree(tree)+geom_text2(aes(subset=!isTip, label=node), hjust=-.3)+geom_tiplab(size=2)

```



```{r, warning=FALSE, message=FALSE}
library(ggrepel)
t1<-ggtree(tree)
t1$data<-separate(t1$data, label, sep="/", into = c("SH-aLRT","UF-Boot"), remove=FALSE)
t1$data$`SH-aLRT`<-as.numeric(t1$data$`SH-aLRT`)
t1$data$`UF-Boot`<-as.numeric(t1$data$`UF-Boot`)
t1$data<-mutate(t1$data, Supported =ifelse(`SH-aLRT` >= 80 & `UF-Boot` >= 95, "YES", "N0"))

t2<-t1+geom_nodepoint(aes(subset = !isTip & Supported =="YES", label=Supported), shape=23, fill="darkgrey", alpha=1.0,
                      size=3) +
  geom_tiplab(size=2)
```


Coloring and labelling clades.

```{r, warning=FALSE, message=FALSE}
#setting an offset
x<-.03

t3<-t2 + # geom_hilight(node=67, fill="steelblue", alpha=.25) + # Highlighting MRRT
  geom_cladelabel(node=57, label="O. m. newberri", fontface="italic", offset=x) +
  geom_cladelabel(node=64, label="O. m. gairdnerii", fontface="italic", offset =x)+
  geom_cladelabel(node=67, label="O. m. stonei", fontface="bold.italic", offset =x)+
  geom_cladelabel(node=74, label="O. m. irideus", fontface="italic", offset =x)+
  geom_cladelabel(node=50, label="O. m. whitei", fontface="italic", offset =x)+
  geom_cladelabel(node=53, label="O. m. gilberti", fontface="italic", offset =x)+
  geom_cladelabel(node=46, label="O. m. aquabonita", fontface="italic", offset =x)+
  geom_cladelabel(node=45, label="Golden Trout\nComplex ", offset =.055, fontsize=3, barsize=0)+
  geom_cladelabel(node=79, label="O. m. aquilarum", fontface="italic", offset =x)+
  xlim(0,.2)+
  ggtitle("A")+
  theme(plot.title = element_text(face="bold", size=18))

t3
```




Figure 2B the consensus tree generated from bootstrap replicates.    

```{r, warning=FALSE, message=FALSE, echo=FALSE}
u<-read.tree(file="Data Supplement/ML tree results/ml.contree")
u<-midpoint(u)
u2<-ggtree(u)
u2$data<-mutate(u2$data, Supported = ifelse(as.numeric(label) > 95, "YES","NO"))
u2$data<-mutate(u2$data, Supported90 = ifelse(as.numeric(label) <= 95 & as.numeric(label) >90, "YES","NO"))

u2 + geom_text2(aes(subset=!isTip, label=node), hjust=-.3)

u3<-u2+geom_nodepoint(aes(subset = !isTip & Supported =="YES", label=Supported), shape=21, fill="black", alpha=1.0,
                      size=3) +
  geom_nodepoint(aes(subset = !isTip & Supported90 =="YES", label=Supported90), shape=21, fill="darkgrey", alpha=1.0,
                      size=3) +
  geom_tiplab(size=2) 
  

```

Coloring and labelling clades.

```{r, warning=FALSE, message=FALSE}
u4<-u3 + # geom_hilight(node=67, fill="steelblue", alpha=.25) + # Highlighting MRRT
  geom_cladelabel(node=64, label="O. m. newberri", fontface="italic", offset=x) +
  geom_cladelabel(node=71, label="O. m. gairdnerii", fontface="italic", offset =x)+
  geom_cladelabel(node=74, label="O. m. calisulat", fontface="bold.italic", offset =x)+
  geom_cladelabel(node=58, label="O. m. irideus", fontface="italic", offset =x)+
  geom_cladelabel(node=51, label="O. m. whitei", fontface="italic", offset =x)+
  geom_cladelabel(node=54, label="O. m. gilberti", fontface="italic", offset =x)+
  geom_cladelabel(node=47, label="O. m. aquabonita", fontface="italic", offset =x)+
  geom_cladelabel(node=46, label="Golden Trout\nComplex ", offset =.055, fontsize=3, barsize=0)+
  geom_cladelabel(node=56, label="O. m. aquilarum", fontface="italic", offset =x)+
  xlim(0,.2)+
  ggtitle("B")+
  theme(plot.title = element_text(face="bold", size=18))


u4
```

Put these two together on one page.
```{r}
library(cowplot)
plot_grid(t3,u4, ncol=2)
```


## Species Tree

```{r, warning=FALSE, message=FALSE, out.width='75%'}
t2<-read.nexus(file="Data Supplement/SVDQuartets/data.asc.tre")
t2<-root(t2, c("LCT"))
ggtree(t2,branch.length = "none")+geom_tiplab(size=4)+geom_nodelab()+xlim(0,8)
```

Improving.      

```{r, warning=FALSE, message=FALSE, echo=FALSE}

tree2<-read.nexus(file="Data Supplement/SVDQuartets/data.asc.tre")
tree2<-root(tree2, c("LCT"))

ggtree(tree2, branch.length = "none")+geom_tiplab(size=2.5)+xlim(0,9)+geom_text2(aes(subset=!isTip, label=node), hjust=-.3)
```

```{r}
y<-.5
tree3<- ggtree(tree2, branch.length = "none")+  
  geom_cladelabel(node=31, label="O. m. newberri", fontface="italic", offset=y) +
  geom_cladelabel(node=34, label="O. m. gairdnerii", fontface="italic", offset =y)+
  geom_cladelabel(node=38, label="O. m. calisulat", fontface="bold.italic", offset =y)+
  geom_cladelabel(node=35, label="O. m. irideus", fontface="italic", offset =y)+
  geom_cladelabel(node=26, label="O. m. whitei", fontface="italic", offset =y)+
  geom_cladelabel(node=27, label="O. m. gilberti", fontface="italic", offset =y)+
  geom_cladelabel(node=24, label="O. m. aquabonita", fontface="italic", offset =y)+
  geom_cladelabel(node=23, label="Golden Trout\nComplex ", offset =3, fontsize=3, barsize=0)+
  geom_cladelabel(node=37, label="O. m. aquilarum", fontface="italic", offset =y)+
  xlim(0,12)+
  geom_tiplab(size=2.5)

tree3

```

#Networks

I'll make a nexus file for analysis in splitstree.
```{r, eval=FALSE, warning=FALSE, message=FALSE}
library(treeio)
data<-read.phylip.seq("phylogenetics/test14-maf05p-sites.sort.prune.reheadered.min4.asc.phy")
write.nexus.data(data, file="phylogenetics/splits.nex")
```


```{r nnet, echo=FALSE, fig.cap="Neighbor-Net", out.width = '100%'}
knitr::include_graphics("Data Supplement/SplitsTree/splits.jpg")
```

#Testing alternative hypotheses against consensus tree from IQ-TREE
Best-fit model: TVM+F+ASC+R2 chosen according to BIC

```{r}

tree<-read.tree(file="phylogenetics/test14-maf05p-sites.sort.prune.reheadered.min4.asc.phy.contree")
tree<-root(tree, c("PCT_PPP1_01","LCT_GACR_10"))
ggtree(tree)+geom_tiplab(size=2)+geom_nodelab()

```

Let's do some testing.    

__1__  Our unconstrained tree
```{sh, eval=FALSE}
iqtree -s example.phy -m TVM+F+ASC+R2 -st DNA --prefix example.unconstr
```

```{r}
tree<-read.tree(file="phylogenetics/testing/example.unconstr.treefile")
tree<-root(tree, c("PCT_PPP1_01","LCT_GACR_10"))
ggtree(tree)+geom_tiplab(size=2)+geom_nodelab()

```

__2__ MRRT and Golden Trout Complex are sister
```{sh, eval=FALSE}
iqtree -s example.phy -m TVM+F+ASC+R2 -st DNA -g constrain1.tre --prefix example.constr1
```

```{r}
tree<-read.tree(file="phylogenetics/testing/example.constr1.treefile")
tree<-root(tree, c("PCT_PPP1_01","LCT_GACR_10"))
ggtree(tree)+geom_tiplab(size=2)+geom_nodelab()
```

__3__ MRRT and redbands are monophyletic
```{sh, eval=FALSE}
iqtree -s example.phy -m TVM+F+ASC+R2 -st DNA -g constrain2.tre --prefix example.constr2 --redo
```

```{r}
tree<-read.tree(file="phylogenetics/testing/example.constr2.treefile")
tree<-root(tree, c("PCT_PPP1_01","LCT_GACR_10"))
ggtree(tree)+geom_tiplab(size=2)+geom_nodelab()
```

__4__ MRRT and Coastal Rainbow Trout are monophyletic
```{sh, eval=FALSE}
iqtree -s example.phy -m TVM+F+ASC+R2 -st DNA -g constrain3.tre --prefix example.constr3 --redo
```

```{r}
tree<-read.tree(file="phylogenetics/testing/example.constr3.treefile")
tree<-root(tree, c("PCT_PPP1_01","LCT_GACR_10"))
ggtree(tree)+geom_tiplab(size=2)+geom_nodelab()
```

# Now we combine our treefiles

```{sh, eval=FALSE}
cat example.unconstr.treefile ../test14-maf05p-sites.sort.reheadered.min4.asc.phy.contree example.constr1.treefile example.constr2.treefile example.constr3.treefile > example.treels

iqtree -s example.phy -z example.treels -te ../test14-maf05p-sites.sort.reheadered.min4.asc.phy.contree  -zb 10000 -zw -au -m TVM+F+ASC+R2 -st DNA -redo

#Also
#iqtree -s example.phy -z example.treels -n 0 -zb 10000 -zw -au -m TVM+F+ASC+R2 -st DNA -redo

```

Tree (1) unconstrained     
Tree (2) consensus tree   
Tree (3) constrained MRRT and CA Golden Trout Complex are monophyletic     
Tree (4) constrained redbands are monophyletic     
Tree (5) constrained MRRT+O.m.irideus are monophyletic      

    171   1 -26508.20531       0   0.631 +  0.733 +      1 +  0.733 +  0.959 +     0.629 +    0.851 +      
    172   2  -26541.6246  33.419  0.0832 +  0.104 +  0.173 +  0.104 +  0.167 +    0.0827 +    0.117 +     
    173   3 -26528.18073  19.975  0.0161 -  0.107 +  0.315 + 0.0644 +  0.171 +    0.0168 -   0.0528 +     
    174   4 -26516.46065  8.2553    0.23 +  0.267 +  0.623 +  0.267 +  0.633 +      0.23 +    0.422 +     
    175   5 -26523.39063  15.185  0.0405 + 0.0747 +  0.418 + 0.0747 +  0.203 +     0.042 +    0.125 +     
       
`iqtree -s example.phy -z example.treels -n 0 -zb 10000 -zw -au -m TVM+F+ASC+R2 -st DNA -redo`   

Using an initial parsimony tree to estimate model parameters instead of user supplied tree.    

    170   1 -26508.08365       0   0.632 +   0.73 +      1 +   0.73 +   0.96 +     0.629 +    0.855 +      
    171   2 -26541.67416  33.591  0.0836 +  0.102 +  0.169 +  0.102 +  0.165 +    0.0835 +     0.11 +     
    172   3 -26528.06953  19.986  0.0145 -  0.106 +  0.315 + 0.0649 +  0.174 +    0.0168 -   0.0599 +     
    173   4 -26516.38659  8.3029   0.227 +   0.27 +   0.63 +   0.27 +  0.638 +     0.227 +    0.425 +     
    174   5 -26523.28594  15.202  0.0433 + 0.0754 +  0.422 + 0.0754 +  0.207 +    0.0444 +    0.112 +     

Same results.    

#Wait, what happened?

Some evidence that MRRT and CA Golden Trout Complex are not monophyletic.      